osgar/drivers/lora.py

```diff
@@ -0,0 +1,100 @@
+  The messages have limited size 40 bytes.
+
+  At the moment driver supports only transmission of "pose2d" messages so the control
+  center (and all neighbor robots) know about each other.
+
+  Current protocol uses numbers for DeviceID (1, 2, ... 5) for source identification
+  and '|' for data separation. The whole packet is terminated by '\n'. The DeviceID
+  is added in front. The message history is collected in reverse order.
+"""
+from datetime import timedelta
+
+from osgar.node import Node
+from osgar.bus import BusShutdownException
+
+ALIVE_MESSAGE = b'alive'
+
+
+def parse_lora_packet(packet):
+    """
+       Parse block of binary data. The block starts with integer
+       numbers interlaced by '|'. List of transmitter IDs is returned
+       with the actual payload data. In the case of missing transmitted
+       history header None is returned.
+    """
+    s = packet.split(b'|')
+    try:
+        return [int(x) for x in s[:-1]], s[-1].strip()
+    except ValueError:
+        print('Invalid LoRa packet:', packet)
+        return None, packet
+
+
+def split_lora_buffer(buf):
+    if b'\n' not in buf:
+        return buf, b''
+    i = buf.index(b'\n')
+    return buf[i+1:], buf[:i+1]
+
+
+class LoRa(Node):
+    def __init__(self, config, bus):
+        super().__init__(config, bus)
+        self.device_id = config.get('device_id')  # None for "autodetect"
+        self.min_transmit_dt = timedelta(seconds=10)  # TODO config?
+        self.buf = b''
+        self.raw = b''  # should be defined by Node
+        self.last_transmit = None
+
+    def send_data(self, data):
+        self.last_transmit = self.publish('raw', data + b'\n')
+        return self.last_transmit
+
+    def update(self):
+        channel = super().update()  # define self.time
+        if channel == 'raw':
+            print(self.time, "update", channel, self.raw)
+            self.buf, packet = split_lora_buffer(self.buf + self.raw)
+            if len(packet) > 0:
+                addr, data = parse_lora_packet(packet)
+                if addr is not None and self.device_id is not None and self.device_id not in addr:
+                    print('re-transmit')
+                    self.send_data(self.raw.strip())  # re-transmit data from other robots
+
+        elif channel == 'pose2d':  # accept all types of messages?
+            if self.last_transmit is None or self.time > self.last_transmit + self.min_transmit_dt:
+                self.send_data(bytes(str(self.pose2d), encoding='ascii'))
+        else:
+            assert False, channel  # not supported
+
+    def run(self):
+        start_time = self.send_data(ALIVE_MESSAGE)
+        try:
+            if self.device_id is None:
+                # generate unique messages so that it can be identified
+                # among received messages. The device writes all transmitted
+                # and received messages. In case of transmitted messages there
+                # is my DeviceID.
```

Comment:
Where does the device writes the messages?

Reply:
> In case of transmitted messages there is my DeviceID.

I don't follow. We are receiving messages that we are sending? Without anyone on the other end? And the DeviceIDs are being prepended by the device?

I am seriously lost here :sob:

Reply:
Connection is via serial, which is handled by OSGAR old module. Lora uses "raw" in and "raw" out.

Reply:
Than much I understood from the config. But are we receiving messages that we are sending? Without anyone on the other end?

Reply:
Who is building the address list? The device itself?

Reply:
The device has its own ID. The messages are broadcasted, so anybody can receive them (including other LoRa devices, which is potential unhandled risk). I added some more comments